Berry/automate character tier management

aquisenberry opened the pull request:
### ❓ Type of change

- [x] ✨ New feature (a non-breaking change that adds functionality)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)

### 📚 Description

<!-- Describe your changes in detail -->
- organize utils.js into a utils directory and break out functionalities into their own files
- replace some WET code with functions to follow a more DRY pattern
  - awardXP command has been updated to have a number of repeated codeblocks written as functions
  - requestXP command has been updated to have a number of repeated codeblocks written as functions
- guild service has been updated to add a number of methods to handle creating tables new that will be added at registration
- command created to force new "register" tables to be added after the fact: `/apply_registration_updates`
- command created to Add/Update/Delete character tiers: `/edit_tiers`
- added tier role evaluation upon character XP update.


<!-- Why is this change required? What problem does it solve? -->
This is a feature request that takes the tier roles out of the hands of hte players and automates it
### 📝 Checklist

- [ ] I'm adding a dependency and have discussed it with the entire team first.
- [x] I've tested this code locally.
- [ ] I have added tests (if possible).
- [ ] I've made changes that should be documented for others (env variables, build scripts, etc). I have updated the readme to reflect these changes.

aquisenberry committed:
organize utils

aquisenberry committed:
add update registration command

aquisenberry committed:
rename command

aquisenberry committed:
add edit character tier command

aquisenberry committed:
do some organizing files; add role management functionality; add level up role automation

aquisenberry committed:
add option to view tiers

iwalsh commented on xpholder/commands/mod/awardXp.js:
```diff
@@ -1,323 +1,360 @@
+          { name: "Set Level", value: "set_level" },
+          { name: "Set XP", value: "set_xp" },
+          { name: "Give XP", value: "give_xp" },
+          { name: "Set CXP", value: "set_cxp" },
```
Not critical, but I think we could comment out / remove these "CXP" options as Enter Ravenloft only uses regular XP. (They're commented out already in `everyone/requestXp.js`)

aquisenberry replied:
Happy to remove, but I offer the counterpoint of this: ER doesn't _currently_ use CXP. I'm actually not familiar with what it is, and/or how it is tied to other systems in the bot. Is it worth doing that as a separate cleanup branch?

iwalsh replied:
Yeah fair enough, we can leave it as a separate task. I'm not 100% sure myself - I'd have to reverse-engineer the meaning from the code, or ask JT! 😅

iwalsh commented on xpholder/commands/mod/awardXp.js:
```diff
@@ -1,323 +1,360 @@
+          { inline: false, name: "Progress", value: progressBar },
+        ];
+        break;
+      case "set_cxp":
```
Likewise here, I think it's safe to remove the CXP code. (We'll have it in git history if we ever care about it in the future.)

aquisenberry replied:
similar thoughts as above

iwalsh commented on xpholder/commands/mod/awardXp.js:
```diff
@@ -1,323 +1,360 @@
+  if (!collectorChannel) {
+    return;
+  }
+  const collector = collectorChannel.createMessageComponentCollector({
```
This isn't something you have to do in this PR, but seeing this reminds me that we ought to change this to be more similar to ` handleRequestXpCommandButton()` in the `requestXP.js` command.

This way of doing it with a MessageComponentCollector is not durable: if the XPHolder process terminates (because the container on Heroku restarted, probably), the collector will be lost and that means that the buttons will become unresponsive in the Discord UI. The alternative is to define a handler function as we do in `requestXp.js`, and then it will persist across restarts.

aquisenberry replied:
agreed on the change and that it is outside the scope of this pr. my formatting settings automate alot of things for readability, and I haven't taken the time to turn them off or set up a formatting commit hook on the project. apologies that there is so much to dig through

iwalsh commented on xpholder/commands/mod/editCharacterTiers.js:
```diff
@@ -0,0 +1,78 @@
+      option
+        .setName("bonus")
+        .setDescription(
+          "Multiplyer To Awareded XP ( negatives remove from database )"
```
Typos: "Multiplier to Awarded XP (-1 to delete the role)"

iwalsh commented on xpholder/commands/mod/editCharacterTiers.js:
```diff
@@ -0,0 +1,78 @@
+      option
+        .setName("maximum_level")
+        .setDescription(
+          "Last level of the tier (must be equal to or larger than minimum_level)"
```
I don't think the parentheses are strictly necessary - probably just "Last level of the tier" is sufficient :)

aquisenberry replied:
definitely not strictly necessary, as I already handle the correction of it on the backend. just adding documentation for the user to know why they may get a different result than they expect if they confuse maximum and minimum

iwalsh commented on xpholder/commands/mod/viewGameRules.js:
```diff
@@ -1,243 +1,401 @@
-    return new EmbedBuilder()
-        .setTitle("Config")
-        .setDescription(`**Level Up Message :**
+function buildCongifEmbed(configObj, tiersArr) {
```
Typo (not yours): "buildConfigEmbed"

iwalsh commented on xpholder/commands/mod/viewGameRules.js:
```diff
@@ -1,243 +1,401 @@
+        name: "Mod Role",
+        value: `<@&${configObj.moderationRoleId}>`,
+      },
+      { inline: true, name: "Role Bonus", value: `${configObj.roleBonus}` },
```
Are the "Role Bonus" and older "Tier 1", "Tier 2" fields now redundant with the changes you're making to allow custom tier definitions? (Can they be removed, or do we need them for backwards compatibility?)

aquisenberry replied:
Short answer, yes they are redundant. This feature attempts to move them to a more flexible way of handling tier roles. In this release, both versions still exist, as I'm not 100% certain of what else the old version of the roles affects and/or what it would take to extract them and maintain all current functionality. At a later point, my intention is to remove the old version.

iwalsh commented on xpholder/commands/mod/viewGameRules.js:
```diff
@@ -1,243 +1,401 @@
+  if (!collectorChannel) {
+    return;
+  }
+  const collector = collectorChannel.createMessageComponentCollector({
```
Same remark here about `createMessageComponentCollector` not persisting across restarts. (It's less important here since there's no "approve" button as for requestXp.)

Maybe I'll try to tackle all of these comprehensively in a cleanup pass at some point.

iwalsh reviewed (commented):
Ok, I've gotten through all of the changes under `commands/` - I'll finish reviewing the rest tomorrow. Thanks for taking on this enhancement + cleanup - it's a big task!

iwalsh reviewed (approved):
Ok I've made it through everything! It's a big diff but the bulk is just reformatting. I think I follow the core addition of the new tier management stuff, so I'm good to proceed.

Once this is merged I'll get it deployed - probably sometime this weekend. I don't think there are any new env variables needed so I'll just take a DB backup and then push it out. Thanks for working on this!

aquisenberry committed:
address pr feedback

aquisenberry commented:
I've addressed comments:

- corrected buildCongifEmbed typo
- corrected multiplyer typo
- removed unnecessary qualifier for tier maximum_level option
